File: arrays/kadane_and_like/min_size_subarray_sum.rb
# 209. Minimum Size Subarray Sum
# https://leetcode.com/problems/minimum-size-subarray-sum/

# class Solution:
#     def minSubArrayLen(self, s, nums):
#         """
#         :type s: int
#         :type nums: List[int]
#         :rtype: int
#         """
#         i = j = curSum = 0
#         mini = math.inf
        
#         while j < len(nums):
#             if curSum + nums[j] < s:
#                 curSum = curSum + nums[j]
#                 j += 1
#             else:
#                 mini = min(j-i+1, mini)
#                 curSum = curSum - nums[i]                   
#                 i += 1
                                          
#         return mini if mini != math.inf else 0

File: sandbox/min_size_subarray_sum.py
class Solution:
    def minSubArrayLen(self, s: int, nums: List[int]) -> int:
        n = len(nums)
        l, r = 0, 0
        min_len, cur_sum = n + 1, 0

        while r < n:
            cur_sum += nums[r]

            while cur_sum >= s:
                min_len = min(min_len, r - l + 1)

                cur_sum -= nums[l]
                l += 1
            r += 1

        return 0 if min_len == n + 1 else min_len

